From 12545270de900566b2ed992f68f67f67017eef17 Mon Sep 17 00:00:00 2001
From: Andy CrossGate Yan <GeForce8800Ultra@gmail.com>
Date: Sun, 13 Jul 2025 12:06:17 +0800
Subject: [PATCH 11/11] TrebleSettings: Disable gamma conversion for brightness
 slider

Functionally similar to "Set alternative brightness curve" in TD-based,
but backed by LineageOS addition
(https://github.com/LineageOS/android_frameworks_base/commit/a9664e0fcc42bdcdd293c94df5587024a3edd907 + persist prop)

Change-Id: Iffe314909db067b256aa034100f67a832599431e
---
 res/values/strings.xml                        |  4 ++
 res/xml/treble_settings.xml                   |  5 ++
 ...ssGammaConversionPreferenceController.java | 53 +++++++++++++++++++
 .../settings/treble/TrebleSettings.java       |  1 +
 4 files changed, 63 insertions(+)
 create mode 100644 src/com/android/settings/treble/DisableBrightnessGammaConversionPreferenceController.java

diff --git a/res/values/strings.xml b/res/values/strings.xml
index 393a530135c..47b3010fe0f 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -6975,6 +6975,10 @@
     <string name="disable_voice_call_in_summary">Reboot required</string>
     <!-- Treble settings screen, screen resolution and refresh rate settings title -->
     <string name="screen_resolution_refresh_rate_title">Screen resolution &amp; refresh rate</string>
+    <!-- Treble settings screen, disable gamma conversion for brightness slider title -->
+    <string name="disable_brightness_gamma_conversion_title">Disable gamma conversion for brightness slider</string>
+    <!-- Treble settings screen, disable gamma conversion for brightness slider summary -->
+    <string name="disable_brightness_gamma_conversion_summary">Reboot required</string>
     <!-- Treble settings screen, use alternative backlight scale title -->
     <string name="use_alternative_backlight_scale_title">Use alternative backlight scale</string>
     <!-- Treble settings screen, override minimum brightness title -->
diff --git a/res/xml/treble_settings.xml b/res/xml/treble_settings.xml
index b8449913dfc..65dc7cbcc6f 100644
--- a/res/xml/treble_settings.xml
+++ b/res/xml/treble_settings.xml
@@ -39,6 +39,11 @@
             android:key="screen_resolution_refresh_rate"
             android:title="@string/screen_resolution_refresh_rate_title" />
 
+        <SwitchPreference
+            android:key="disable_brightness_gamma_conversion"
+            android:title="@string/disable_brightness_gamma_conversion_title"
+            android:summary="@string/disable_brightness_gamma_conversion_summary" />
+
         <SwitchPreference
             android:key="use_alternative_backlight_scale"
             android:title="@string/use_alternative_backlight_scale_title" />
diff --git a/src/com/android/settings/treble/DisableBrightnessGammaConversionPreferenceController.java b/src/com/android/settings/treble/DisableBrightnessGammaConversionPreferenceController.java
new file mode 100644
index 00000000000..c33d6ec8e1e
--- /dev/null
+++ b/src/com/android/settings/treble/DisableBrightnessGammaConversionPreferenceController.java
@@ -0,0 +1,53 @@
+package com.android.settings.treble;
+
+import android.content.Context;
+import android.os.SystemProperties;
+
+import androidx.preference.Preference;
+import androidx.preference.PreferenceScreen;
+import androidx.preference.SwitchPreference;
+
+import com.android.settings.core.BasePreferenceController;
+
+public class DisableBrightnessGammaConversionPreferenceController extends BasePreferenceController
+        implements Preference.OnPreferenceChangeListener {
+
+    private static final String DISABLE_BRIGHTNESS_GAMMA_CONVERSION_KEY = "disable_brightness_gamma_conversion";
+    private static final String TREBLE_DISABLE_BRIGHTNESS_GAMMA_CONVERSION_PROPERTY = "persist.sys.brightness.disable_gamma_conversion";
+
+    private SwitchPreference mSwitchPreference;
+
+    public DisableBrightnessGammaConversionPreferenceController(Context context) {
+        super(context, DISABLE_BRIGHTNESS_GAMMA_CONVERSION_KEY);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return AVAILABLE;
+    }
+
+    @Override
+    public String getPreferenceKey() {
+        return DISABLE_BRIGHTNESS_GAMMA_CONVERSION_KEY;
+    }
+
+    @Override
+    public void displayPreference(PreferenceScreen screen) {
+        mSwitchPreference = screen.findPreference(getPreferenceKey());
+        super.displayPreference(screen);
+    }
+
+    @Override
+    public void updateState(Preference preference) {
+        boolean checked = SystemProperties.getBoolean(TREBLE_DISABLE_BRIGHTNESS_GAMMA_CONVERSION_PROPERTY, false);
+        mSwitchPreference.setChecked(checked);
+    }
+
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+        SystemProperties.set(TREBLE_DISABLE_BRIGHTNESS_GAMMA_CONVERSION_PROPERTY, String.valueOf((boolean) newValue));
+        mSwitchPreference.setChecked((boolean) newValue);
+        return true;
+    }
+
+}
diff --git a/src/com/android/settings/treble/TrebleSettings.java b/src/com/android/settings/treble/TrebleSettings.java
index b4c170773df..100dc7f8286 100644
--- a/src/com/android/settings/treble/TrebleSettings.java
+++ b/src/com/android/settings/treble/TrebleSettings.java
@@ -38,6 +38,7 @@ public class TrebleSettings extends DashboardFragment {
         controllers.add(new UseAlternativeAudioJackDetectionPreferenceController(context));
         controllers.add(new DisableVoiceCallInPreferenceController(context));
         controllers.add(new ScreenResolutionRefreshRatePreferenceController(context));
+        controllers.add(new DisableBrightnessGammaConversionPreferenceController(context));
         controllers.add(new UseAlternativeBacklightScalePreferenceController(context));
         controllers.add(new OverrideMinimumBrightnessPreferenceController(context));
         controllers.add(new DisableSfHwcBackpressurePreferenceController(context));
-- 
2.43.0

